File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
    int depth(TreeNode* node, int& maxDepth) {
        if (!node)
            return 0;
        auto left = depth(node->left, maxDepth);
        auto right = depth(node->right, maxDepth);
        maxDepth = std::max(maxDepth, left + right);

        return max(left, right) + 1; // return max tree depth
    }
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int maxDepth = 0;
        depth(root, maxDepth);
        return maxDepth;
    }
};

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        self.max_diameter = 0

        def depth(node):
            if not node:
                return 0
            left = depth(node.left)
            right = depth(node.right)
            # aktualizujemy maksymalną średnicę przy przechodzeniu przez ten węzeł
            self.max_diameter = max(self.max_diameter, left + right)
            # zwracamy wysokość drzewa
            return max(left, right) + 1

        depth(root)
        return self.max_diameter